### ISSUE TITLE
PKCS11 function generate keypair not supported by Gemalto IAS-ECC

### ISSUE THREAD

**Issue by DBarthe** [OpenSC/OpenSC#2056]
### Problem Description

I'm trying to add support for OpenSC to a card management system (generating keypair, storing certificates, etc). IAS-ECC is one kind of card that is used within the system. But when testing with some Gemalto IAS ECC cards, it fails to generate a keypair, while this is working well with other proprietary IAS-ECC drivers.

### Steps to reproduce

Using the CLI on windows.

The slot listing gives me 3 slots for one card inserted... I used 0x2 because it is writable. I never had all these choices using other pkcs11 drivers :
```
> pkcs11-tool  -L
Available slots:
Slot 0 (0x0): Gemalto Prox-DU Contact_10500170 0
  token label        : ECC eID (Card PIN)
  token manufacturer : Gemalto
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized, readonly
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         :
  pin min/max        : 4/4
Slot 1 (0x1): Gemalto Prox-DU Contact_10500170 0
  token label        : ECC eID (Signature PIN)
  token manufacturer : Gemalto
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized, readonly
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         :
  pin min/max        : 6/6
Slot 2 (0x2): Gemalto Prox-DU Contact_10500170 0
  token label        : ECC eID (Card PIN)
  token manufacturer : Gemalto
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         :
  pin min/max        : 4/4
```

Then generating the keypair :

```
> pkcs11-tool.exe  --slot 0x02 --login --pin *  --keypairgen
error: PKCS11 function C_GenerateKeyPair failed: rv = CKR_FUNCTION_NOT_SUPPORTED (0x54)
Aborting.
```
Then if I list object on the card, I see a private key has been generated :
```
Private Key Object; RSA
  label:      Private Key
  ID:         0edbeb8670f3ad3bffef2f042a2cf9afb1981b02
  Usage:      sign
  Access:     sensitive, always sensitive, never extractable, local
```

### Logs
The most interesting part I guess (it's huge otherwise):
```
P:13384; T:4496 2020-06-10 18:38:54.244 [opensc-pkcs11] pkcs15-iasecc.c:1825:iasecc_emu_store_data: called
P:13384; T:4496 2020-06-10 18:38:54.244 [opensc-pkcs11] pkcs15-iasecc.c:1613:iasecc_store_prvkey: called
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] Private Key id '0edbeb8670f3ad3bffef2f042a2cf9afb1981b02'
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] MD style '0x1'
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] pkcs15-iasecc.c:1530:iasecc_md_gemalto_new_prvkey: called
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] pkcs15.c:2787:sc_pkcs15_get_object_guid: called
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] pkcs15.c:2865:sc_pkcs15_aux_get_md_guid: called
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] pkcs15.c:2874:sc_pkcs15_aux_get_md_guid: returning with: -1408 (Not supported)
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] card.c:1014:sc_card_ctl: called
P:13384; T:4496 2020-06-10 18:38:54.245 [opensc-pkcs11] card-iasecc.c:2539:iasecc_get_serialnr: returning with: -1408 (Not supported)
P:13384; T:4496 2020-06-10 18:38:54.246 [opensc-pkcs11] card_ctl(5) not supported
P:13384; T:4496 2020-06-10 18:38:54.246 [opensc-pkcs11] pkcs15.c:2824:sc_pkcs15_get_object_guid: 'GET_SERIALNR' CTL failed and other serial numbers not present: -1408 (Not supported)
P:13384; T:4496 2020-06-10 18:38:54.246 [opensc-pkcs11] pkcs15-iasecc.c:1536:iasecc_md_gemalto_new_prvkey: Cannot get private key GUID: -1408 (Not supported)
P:13384; T:4496 2020-06-10 18:38:54.246 [opensc-pkcs11] pkcs15-iasecc.c:1622:iasecc_store_prvkey: MD: cannot add new key: -1408 (Not supported)
P:13384; T:4496 2020-06-10 18:38:54.246 [opensc-pkcs11] pkcs15-iasecc.c:1845:iasecc_emu_store_data: returning with: -1408 (Not supported)
P:13384; T:4496 2020-06-10 18:38:54.246 [opensc-pkcs11] pkcs15-lib.c:1559:sc_pkcs15init_generate_key: Card specific 'store data' failed: -1408 (Not supported)
P:13384; T:4496 2020-06-10 18:38:54.246 [opensc-pkcs11] sc_pkcs15init_generate_key returned -1408
P:13384; T:4496 2020-06-10 18:38:54.247 [opensc-pkcs11] libopensc return value: -1408 (Not supported)
```

Investigating the source code, it seems to come from [pkcs15.c#L2833](https://github.com/OpenSC/OpenSC/blob/7ae74c524f1cc2912e8de1ecbf616b602424a830/src/libopensc/pkcs15.c#L2833) :
the serial number of the card can not be obtained from the tokenInfo.

I don't know what is happening. Is there anyone aware of this kind of issue?

**Comment by DBarthe** [comment]
Hi @Jakuje,
Sounds like a great task for newcomer like me, I'll give a try sooner or later.

**Comment by Jakuje** [comment]
Yes. Looking forward to a pull request!

**Comment by misterzed88** [comment]
This is now fixed in master in c903ddfce1279fc5023613b232e9b5b07d93c766.

**Comment by Jakuje** [comment]
Closing as it is merged now in master via #2070
